File: week2/job_queue/job_queue.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using std::vector;
using std::cin;
using std::cout;
using std::swap;
using std::pair;
using std::make_pair;

// shorten the pair<> type
typedef pair<int, long long> pairType;

class Heap {
public:
    vector<pairType> data_;
    int threadCount_ = 0;

    Heap(int threadCount) {
        threadCount_ = threadCount;
        for (int i = 0; i < threadCount; i++)
            data_.push_back(make_pair(i, 0));
    }

    bool comPair(pairType p1, pairType p2){
        if (p1.second != p2.second)
            return p1.second < p2.second;
        else
            return p1.first < p2.first;
    }

    void siftUp(int i) {
        while ((i > 1) && comPair(data_[i], data_[(i-1)/2])) {
            swap(data_[(i-1)/2], data_[i]);
            i = (i-1)/2;
        }
    }

    void siftDown(int i) {
        int maxIndex = i;
        int left = 2*i+1;
        if ((left < threadCount_) &&  comPair(data_[left], data_[maxIndex]))
            maxIndex = left;

        int right = (2*i)+2;
        if ((right < threadCount_) && comPair(data_[right], data_[maxIndex]))
            maxIndex = right;

        if (i != maxIndex) {
            swap(data_[i], data_[maxIndex]);
            siftDown(maxIndex);
        }
    }

    void changePriority(int i, long long p) {
        long long oldP = data_[i].second;
        data_[i] = make_pair(data_[i].first, p);
        if (p > oldP)
            siftUp(i);
        else
            siftDown(i);
        siftDown(i);
    }

    //    void GenerateSwapsMinHeap() {
    //        data_.clear();
    //        for (int x = (threadCount_/2); x >= 0 ; x--) {siftDown(x);}
    //    }
};

class JobQueue {
private:
    int num_workers_;
    vector<int> jobs_;
    vector<int> assigned_workers_;
    vector<long long> start_times_;

    void WriteResponse() const {
        for (int i = 0; i < jobs_.size(); ++i) {
            cout << assigned_workers_[i] << " " << start_times_[i] << "\n";
        }
    }

    void ReadData() {
        int m;
        cin >> num_workers_ >> m;
        jobs_.resize(m);
        for(int i = 0; i < m; ++i)
            cin >> jobs_[i];
    }

    void AssignJobs() {
        assigned_workers_.resize(jobs_.size());
        start_times_.resize(jobs_.size());
        Heap heap = Heap(num_workers_);
        for (int i = 0; i < jobs_.size(); i++) {
            assigned_workers_[i] = heap.data_[0].first;
            start_times_[i] = heap.data_[0].second;
            heap.changePriority(0, heap.data_[0].second + jobs_[i]);
        }
    }

public:
    void Solve() {
        ReadData();
        AssignJobs();
        WriteResponse();
    }
};

int main() {
    std::ios_base::sync_with_stdio(false);
    JobQueue job_queue;
    job_queue.Solve();
    return 0;
}

File: week2/job_queue/job_queue.py
# python3

from collections import namedtuple

AssignedJob = namedtuple("AssignedJob", ["worker", "started_at"])


def assign_jobs(n_workers, jobs):
    # TODO: replace this code with a faster algorithm.
    result = []
    next_free_time = [0] * n_workers
    for job in jobs:
        next_worker = min(range(n_workers), key=lambda w: next_free_time[w])
        result.append(AssignedJob(next_worker, next_free_time[next_worker]))
        next_free_time[next_worker] += job

    return result


def compare(job1: AssignedJob, job2: AssignedJob):
    if job1.started_at != job2.started_at:
        return job1.started_at < job2.started_at
    else:
        return job1.worker < job2.worker


class Heap:
    def __init__(self, n_workers=0):
        self.n_workers = n_workers
        # give each thread a job by setting its start_time to 0
        self.threads = [AssignedJob(i, 0) for i in range(self.n_workers)]

    def sift_up(self, i):
        while (i > 1) and compare(self.threads[i], self.threads[(i - 1) / 2]):
            # swap with its parent if it is smaller than parent
            self.threads[(i-1) / 2], self.threads[i] = self.threads[i], self.threads[(i-1) / 2]
            i = (i-1) / 2

    def sift_down(self, i):
        max_index = i
        left = int(2 * i) + 1

        if (left < self.n_workers) and compare(self.threads[left], self.threads[max_index]):
            max_index = left

        right = int(2 * i) + 2
        if (right < self.n_workers) and compare(self.threads[right], self.threads[max_index]):
            max_index = right

        if i != max_index:
            self.threads[i], self.threads[max_index] = self.threads[max_index], self.threads[i]
            self.sift_down(max_index)

    def change_priority(self, i, p_new):
        old_p = self.threads[i].started_at
        self.threads[i] = AssignedJob(self.threads[i].worker, p_new)
        if p_new > old_p:
            self.sift_up(i)
        else:
            self.sift_down(i)
        self.sift_down(i)


class JobQueue:
    def __init__(self):
        self.n_workers = 0
        self.jobs = []
        self.threads: [AssignedJob] = []

    def read(self):
        self.n_workers, m = map(int, input().split())
        self.jobs = [0] * m
        self.jobs = list(map(int, input().split()))
        assert len(self.jobs) == m

    def write_response(self):
        for job in self.threads:
            print(job.worker, job.started_at)
    '''
    2 5
    1 2 3 4 5
    # there are 2 threads or workers and 5 jobs
    0 0
    1 0
    0 1
    1 2
    0 4
    '''
    def assign_jobs(self):
        self.threads = [AssignedJob(0, 0)] * len(self.jobs)
        heap = Heap(self.n_workers)
        for i in range(len(self.jobs)):
            self.threads[i] = AssignedJob(heap.threads[0].worker, heap.threads[0].started_at)
            # add new job
            heap.change_priority(0, heap.threads[0].started_at + self.jobs[i])


def main():
    queue = JobQueue()
    queue.read()
    queue.assign_jobs()
    queue.write_response()


if __name__ == "__main__":
    main()
